backend/danswer/connectors/sharepoint/connector.py

```diff
@@ -0,0 +1,230 @@
+        
+        return pruned_driveitem_object_list
+    
+    def get_all_driveitem_objects(self, site_object_list: list[Site]) -> list[DriveItem]:
+        driveitem_list = []
+        for site_object in site_object_list:
+            site_list_objects = site_object.lists.get().execute_query()
+            for site_list_object in site_list_objects:
+                try:
+                    driveitems = site_list_object.drive.root.get_files(True).execute_query()
+                    driveitem_list.extend(driveitems)
+                except Exception:
+                    pass
+                
+        return driveitem_list
+
+    def get_all_site_objects(self) -> list[Site]:
+        site_object_list: list[Site] = []
+        
+        sites_object = self.graph_client.sites.get().execute_query()
+
+        if len(self.requested_site_list)>0:
+            for requested_site in self.requested_site_list:
+                adjusted_string = "/" + requested_site
+                for site_object in sites_object:
+                    if site_object.web_url.endswith(adjusted_string):
+                        site_object_list.append(site_object)
+        else:
+            site_object_list.extend(sites_object)
+        
+        return site_object_list
+    
+    def _fetch_from_sharepoint(
+            self, start: datetime | None = None, end: datetime | None = None
+    ) -> GenerateDocumentsOutput:
+        if self.graph_client is None:
+            raise ConnectorMissingCredentialError("Sharepoint")
+        
+        site_object_list = self.get_all_site_objects()
+
+        driveitem_list = self.get_all_driveitem_objects(site_object_list)
+
+        if start is not None and end is not None:
+            driveitem_list = self._prune_driveitem_list_by_time(start, end, driveitem_list)
+
+        # goes over all urls, converts them into Document objects and then yeilds them in batches
+        doc_batch: list[Document] = []
+        batch_count = 0
+        for driveitem_object in driveitem_list:
+
+            doc_batch.append(self.convert_driveitem_object_to_document(driveitem_object))
+
+            batch_count+=1
+            if batch_count>=self.batch_size:
+                yield doc_batch
+                batch_count = 0
+                doc_batch: list[Document] = []
+        yield doc_batch
+    
+    def convert_driveitem_object_to_document(
+            self, driveitem_object:DriveItem, 
+    ) -> Document:
+        file_text = self.extract_driveitem_text(driveitem_object)
+        doc = Document(
+            id=driveitem_object.id, 
+            sections=[Section(link=driveitem_object.web_url, text=file_text)],
+            source=DocumentSource.SHAREPOINT,
+            semantic_identifier=driveitem_object.name,
+            doc_updated_at=driveitem_object.last_modified_datetime.replace(tzinfo=timezone.utc), 
+            primary_owners=[BasicExpertInfo(display_name=driveitem_object.last_modified_by.user.displayName,
```

Comment:
Is this guaranteed to exist?

Reply:
pretty certain a document always has a user